Refactor volume test in a similar way to csi tests

mkimuram opened the pull request:
Refactoring for non-csi e2e test similar to below commit in csi e2e test.
https://github.com/kubernetes/kubernetes/commit/4d11dab272fc358c918a1a6b83db8f938535cdb0#diff-0d9ecaa3e6a0297186ad33f57aad472e

Scopes for this refactoring are below three files:
  - test/e2e/storage/volumes.go
  - test/e2e/storage/volume_io.go
  - test/e2e/storage/persistent_volumes-volumemode.go

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #66571

**Special notes for your reviewer**:
/sig storage
@davidz627 @msau42 @jsafrane  

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

neolit123 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,709 @@
+
+////////////////////////////////////////////////////////////////////////
+// NFS
+////////////////////////////////////////////////////////////////////////
```
not sure we have comment blocks like that in the repo, but if sig-storage are fine with that. :) :+1:

mkimuram replied:
Thank you for your comments. Actually, they are coming from existing codes. I'm ready to change it, if required.

neolit123 commented:
/ok-to-test

jsafrane commented:
Thanks for starting cleaning up our tests. May I suggest you to include also subpath tests in test/e2e/storage/subpath.go? It already has a mini framework to start/stop storage servers and create various VolumeSpecs and it would be better to have one framework that's used everywhere.

mkimuram commented:
/retest

neolit123 commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-e2e-gce

mkimuram commented:
@neolit123 @jsafrane 

Now, subpath tests are included in this refactoring. PTAL

jsafrane commented on test/e2e/framework/volume_util.go:
```diff
@@ -356,6 +356,34 @@ func StartVolumeServer(client clientset.Interface, config VolumeTestConfig) *v1.
+	Logf("Deleting server pod %q...", serverPod.Name)
+	err := DeletePodWithWait(f, cs, serverPod)
+	if secErr != nil || err != nil {
+		if secErr != nil {
```
Please move to `cs.CoreV1().Secrets(ns.Name).Delete()` and leave only `if err != nil` here. I'd also suggest using `err` instead of `secErr` everywhere.

jsafrane commented on test/e2e/storage/volumes.go:
```diff
@@ -59,6 +59,65 @@ import (
+	// emtpydir doesn't provide persistency, therefore it should not be tested.
+}
+
+var needInjectHTML = map[string]bool{
```
This should be part of `driverInfo`.

mkimuram replied:
I think that needInjectHTML is volumes.go's test specific configuration because whether injectHTML should be called in the tests is up to the tests themselves, therefore I prefer to keep it in test side, not driver side. I'm thinking about refactoring a bit more on each tests, so let's discuss it more after I refactor it.

jsafrane commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,758 @@
+				tests := createVolumeTest(driver)
+				for _, t := range tests {
+					defer framework.VolumeTestCleanup(f, t.config)
+					needInjectHTML, ok := needInjectHTML[t.name]
```
needInjectHTML should be part of `driverInfo`.

jsafrane commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,758 @@
+					driver.cleanupDriver()
+				})
+
+				testSubPath(&testInputs)
```
It seems to me that subpath tests are run twice, subpath.go is the second place.

mkimuram replied:
I put both PV tests and volume tests in the same loop, so testSubPath will be called twice for them.

In other word, I generalize what was done in nfs and nfsPVC and apply it to all other drivers.
So, drivers which return both PV and volume, like nfs, cephFS, rbd, and iSCSI, are tested for both.
Also, tests for drivers which don't return either of them will be skipped.

jsafrane commented on test/e2e/storage/subpath.go:
```diff
@@ -671,131 +813,6 @@ func podContainerExec(pod *v1.Pod, containerIndex int, bashExec string) (string,
-func (s *emptydirSource) cleanupVolume(f *framework.Framework) {
-}
-
 type gcepdPVCSource struct {
```
Please move this GCE PD to non_csi_volumes.go too, we'll add cinder and AWS later.

mkimuram replied:
I will try it. However, it seems that current initGCEPDPartition is always skipped due to the fact that it requires manual operations. So, I'm not sure that I can test my code properly.

msau42 replied:
that's fine, I can help validate gce pd as needed

jsafrane commented on test/e2e/storage/volumes.go:
```diff
@@ -59,6 +59,65 @@ import (
+	tests   []framework.VolumeTest
+}
+
+var expectedContents = map[string]string{
```
This should be part of driverInfo.

jsafrane commented on test/e2e/storage/volumes.go:
```diff
@@ -94,167 +153,6 @@ var _ = utils.SIGDescribe("Volumes", func() {
-	})
-
 	////////////////////////////////////////////////////////////////////////
 	// OpenStack Cinder
```
Please move OpenStack, GCE PD, vSphere and Azure to the framework too and skip them if it's not running on the right cloud.

jsafrane reviewed (commented):
Thanks for the refactoring, it looks better now. Please add also remaining volume types like GCE PD, Cinder and vSpehere to non_csi_volumes.go so we have all volumes on the same spot and all tests use the same framework.

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,758 @@
+				driver.cleanupDriver()
+			})
+
+			It("should be mountable", func() {
```
Can we hide the actual test cases in the suite-specific test file (like what is being done for subpath)?

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,758 @@
+				driver testDriver
+			)
+
+			BeforeEach(func() {
```
I think both subpath and these tests can share the driver setup and cleanup

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,758 @@
+			It("should show a volume with proper volume mode to a pod [Feature:BlockVolume]", func() {
+				tests := createPersistentVolumeModeTest(driver)
+				for _, t := range tests {
+					testVolumeMode(f, cs, t.config, ns.Name, t.scName, t.volBindMode, t.volMode, t.pvSource, t.isBlockSupported)
```
I think each test should still have its own It().  Same for all the tests.

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,758 @@
+					driver = curInitDriver(f, config, node)
+					driver.createDriver()
+
+					testResource = createSubPathTestResource(f, currentTest, driver)
```
can this subpath-specific setup/cleanup be done inside testSubPath?  Ideally, this file should just be a generic test driver and we can keep test-specific logic in separate files.

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1170 @@
+	}
+
+	if r.volSource == nil {
+		framework.Skipf("Driver %s doesn't support PV: %v -- skipping", driverInfo.name, testPattern.isPVtest)
```
it could be that the driver doesn't support PV or inline

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1170 @@
+}
+
+func (g *gceDriver) getPersistentVolumeSource(readOnly bool) *v1.PersistentVolumeSource {
+	return nil
```
we primarily want to test gce pd with PV.  You can copy the GCE PD PV setup from the subpath test.

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1170 @@
+}
+
+type testPattern struct {
+	isPVtest    bool
```
these parameters are not used by all tests.  Can each test suite implement their own type that implements an Interface?  Then you can store test-specific parameters in the specific struct.

mkimuram commented:
/retest

mkimuram commented:
/test pull-kubernetes-e2e-kops-aws

mkimuram commented:
@msau42
Could you help debuging below two gce specific issues?

1. Dynamic PV ([sig-storage] Non-CSI Driver: gce, Test: subPath, TestPattern: Dynamic PV )
   Below test fails for Dynamic PV on gce (and aws), while it passes for NFS Dynamic PV in aws.
   So, I guess that it is cloud provider specific issue.
   (Or I might misunderstand subpath test for readOnly.)

   - should fail for new directories when readOnly specified in the volumeSource

   Note that I borrowed the code from [here](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/subpath.go#L838) and [here](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/volume_provisioning.go#L252)

   
2. Static PV (([sig-storage] Non-CSI Driver: gce, Test: subPath, TestPattern: Static PV ))
   GCE PV test seems to fail on ```detachAndDeletePDs``` called from ```cleanupDriver```
   ([code](https://github.com/kubernetes/kubernetes/blob/f663862bd5b86831355a7fdce368fae129efb7cc/test/e2e/storage/non_csi_volumes.go#L1257))
( ```getVolumeSource```, ```createDriver```, and ```cleanupDriver``` are commented out, for now. I can enable them and re-commit them to make CI test show the test results, if you need it. )

   The cause seems that the volume wasn't found in the node and failed to detach the volume from node.  

   Note that I borrowed the code from [here](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/volumes.go#L535) and [here](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/subpath.go#L922)

Also, could you arrange testing for cloud providers that aren't automatically run on CI tests?
(Or can we merge this PR without testing for such cloud provider and fix bugs once they found,  because they won't be a blocker for CI?)

@msau42 @jsafrane 
Could you review the codes again, please?
I hope that I changed in a way that you are intended in your comments.

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1493 @@
+func skipTestForDriver(testPattern testPattern, driver testDriver) {
+	driverInfo := driver.getDriverInfo()
+	// Check if testVolType is supported by driver
+	if testPattern.testVolType == inlineVolume {
```
might be better to do a switch statement, and error on the default case

mkimuram replied:
Thank you for your advice. I will fix it.

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1493 @@
+func (g *gceDriver) cleanupDriver() {
+	config := g.driverInfo.config
+	// The volume would be mounted on config.clientNodeName
+	detachAndDeletePDs(g.volumeName, []types.NodeName{types.NodeName(config.ClientNodeName)})
```
I don't think the cleanup should be doing detach.

Usually in a test, when the pod is terminated, then the volume automatically gets detached by Kubernetes.

mkimuram replied:
I'm not sure why but it seems to work well, now. 
I will ask you for your help again, once the issue is reproduced.

msau42 commented on test/e2e/storage/subpath.go:
```diff
@@ -177,194 +169,462 @@ var _ = utils.SIGDescribe("Subpath", func() {
+				s.roVolSource = &v1.VolumeSource{
+					PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+						ClaimName: s.pvc.Name,
+						ReadOnly:  true,
```
The readonly subpath tests may need to be restructured a little bit.  The problem with readOnly (especially with dynamic provisioning), is that you need to create the volume first, mount it writable so that it will format a fs on it, write some data to it, and then create a new pod and change it to read only.

mkimuram replied:
I think that the [latest commit](https://github.com/kubernetes/kubernetes/pull/66577/commits/2752097b35fe444f6545cb9aec18c6777365150a) is doing it and I hope that it works well.
(NFS doesn't need formatting, so its test seems working well without this code, which made me confused.)

mkimuram commented:
@msau42 

Now pull-kubernetes-e2e-gce test passed, after I rerun it
(aws test still fails with time out error.)
Does it mean that tests themselves are good but it takes too much time to execute all of them?

I would appreciate it if you could give me any suggestions for the next step. 
(Deleting test patterns for now or add filter in test execution logic, etc ...)

mkimuram commented:
/test pull-kubernetes-e2e-kops-aws

mkimuram commented:
/test pull-kubernetes-e2e-kops-aws

mkimuram commented:
/test pull-kubernetes-e2e-kops-aws

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+
+type inlineVolumeTestDriver interface {
+	testDriver
+	getVolumeSource(readOnly bool, fsType string) *v1.VolumeSource
```
I think for inline volume and static pv interfaces, there needs to be one more call to createVolume, and deleteVolume, instead of being done as part of create/delete driver.  Because for the dynamic provisioning case, you shouldn't need to manually create the volume

mkimuram replied:
Separated them as suggested.

For drivers tested with testServer pod, like NFS, iSCSI, and Ceph, volumes are predefined in container images. So, they can't separate ```createDriver``` logic and ```createVolume``` logic.
In addition, dynamic provision pod for NFS seems to be developed separately from that for staticPV/inlineVolume, therefore a different pod from a different container image needs to be launched.  
(This means that two pods, for staticPV/inlineVolume and dynamicPV, will be launched for NFS in dynamic provision test, if we put testServer creation logic in ```createDriver```.)

Therefore, I put testServer creation logic for these drivers in ```createVolume```, not in ```createDriver```.
(I think testServer creation for these drivers include volume creation, so it would be acceptable.)

Note that most of in-tree drivers do nothing in ```createDriver```, however all csi-drivers will need it to create their driver pods, so I think that it makes sence to keep them separate for them.

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+
+	// PD will be created in framework.TestContext.CloudConfig.Zone zone,
+	// so pods should be also scheduled there.
+	config.NodeSelector = map[string]string{
```
This should be set at the time of inline.createVolume().  PVs (both manual and dynamic) should not set this.

mkimuram replied:
Moved as suggested.

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+
+type driverInfo struct {
+	name                      string
+	fsGroup                   *int64
```
I don't think drivers should set fsgroup.  They may need a supportFsGroup setting though

mkimuram replied:
Defined isFsGroupSupported and use it in testSuite to set fsGroup.

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+	supportedFsType  map[string]bool
+	isBlockSupported bool
+
+	// volumes test specific parameters
```
why do these two parameters need to be set by the driver?  Can you generate a "expected content" randomly based off of test namespace?

mkimuram replied:
Make testSuite always generate content.

My intention was to keep compatibility from the exsiting tests, which prepare volumes with contents.
However, it won't be a must and I agree to change it as suggested to make ```driverInfo``` simple and clean.

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+	needInjectHTML  bool
+
+	// volumeIO test specific parameter
+	testFile string
```
same, can you generate the testFile name based off the test namespace?

mkimuram replied:
It was used to check whether volumes test should be run.
For emptydir, volumes test should not be run for it doesn't provide persistency.

To fix it, I defined ```isPersistent``` in ```driverInfo``` and ```skipUnsupportedTest``` in ```testSuite```.
By doing so, volumes test can be skipped if driver doesn't provide persistency.

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+type driverInfo struct {
+	name                      string
+	fsGroup                   *int64
+	podSec                    v1.PodSecurityContext
```
don't think drivers should be providing pod security context either

mkimuram replied:
Removed it and made it generate inside testSuite by using ```isFsGroupSupported```.

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+				"ext4": true,
+				"xfs":  true,
+			},
+			isBlockSupported: false,
```
it is supported

mkimuram replied:
Fixed for gce. Also fixed for other drivers.

msau42 commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+			expectedContent:  "Hello from GCE from namespace " + ns.Name,
+			needInjectHTML:   true,
+			testFile:         "gce_io_test",
+			f:                f,
```
Can the test framework pass `f`, `config`, and `node` in instead of requiring every driver to set it?

mkimuram replied:
Defined ```setCommonDriverParameters``` and call it in ```BeforeEach```.

msau42 commented on test/e2e/storage/subpath.go:
```diff
@@ -168,203 +137,372 @@ var _ = utils.SIGDescribe("Subpath", func() {
+
+	skipCreatingResourceForProvider(testPattern)
+
+	switch testPattern.testVolType {
```
I see this is already being done [here](https://github.com/kubernetes/kubernetes/pull/66577/files#diff-b2926bcb65edc51156080c46e7e5fcf1R320).  Do we need to repeat it again?  Or can we try to have common wrappers for the shared parts?

mkimuram replied:
```skipCreatingResourceForProvider``` is removed from ```setup```.
Necessary skip will be done ```skipUnsupportedTest``` in test loop, by using ```testDriver```s' and ```testSuite```s' ```skipUnsupportedTest```.

mkimuram commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+			curTestPatternName := testPatternName
+			curTestPattern := testPattern
+
+			for driverName, initDriver := range testDrivers {
```
Moved as suggested.

mkimuram commented on test/e2e/storage/persistent_volumes-volumemode.go:
```diff
@@ -36,6 +35,301 @@ const (
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Creating pv and pvc")
+		switch t.testVolType {
```
I changed testSuite's ```getTestFunc``` to take ```testPattern``` and ```testDriver``` as arguments, and allow ```getTestFunc``` to return a suitable testFunction for the comibination of arguments. I think that this reduces complexity of the test case logic enough and we will be able to apply it to other test suites in the future.

mkimuram commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+	"aws":                        initAwsDriver,
+}
+
+var testCases = map[string]func() testCase{
```
Changed as suggested.

mkimuram commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+	}
+}
+
+// NFS
```
Separated and Renamed as below:

- in_tree_drivers.go: Driver methods only
- in_tree_volumes.go: Rest of logics and interface definitions

mkimuram commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+func skipCreatingResourceForProvider(testPattern testPattern) {
+	fsType := testPattern.fsType
+
+	if fsType == "xfs" && framework.ProviderIs("gce", "gke") {
```
Defined ```skipUnsupportedTest``` in ```testDriver``` and make it called from ```skipUnsupportedTest``` in test loop.

```skipUnsupportedTest``` become responsible for skipping other than FsType, like provider.
(```framework.SkipUnlessProviderIs``` is now moved from ```initDriver```, so it can be put outside BeforeEach.)

mkimuram commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+		config = framework.VolumeTestConfig{
+			Namespace:      ns.Name,
+			Prefix:         "volume",
+			ClientNodeName: node.Name,
```
Removed it from there and only put it in hostpath and hostpathSymlink's ```createVolume```,
because hostpath is not persistent across node and it should always be scheduled on the same node.

We might need to add ```isPersistentAcrossNode``` to ```driverInfo``` and skip topology related tests for drivers like hostpath in the future.

mkimuram commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+			Namespace:      ns.Name,
+			Prefix:         "volume",
+			ClientNodeName: node.Name,
+			ServerNodeName: node.Name,
```
Removed.

mkimuram commented on test/e2e/storage/non_csi_volumes.go:
```diff
@@ -0,0 +1,1567 @@
+	"subPath":                          initSubPathTestCase,
+}
+
+var _ = utils.SIGDescribe("Non-CSI", func() {
```
Changed as suggested.

mkimuram commented:
Note that tests that use RBD image fail currently, due to https://github.com/kubernetes/kubernetes/issues/67269 .

msau42 commented on test/e2e/storage/volume_io.go:
```diff
@@ -61,6 +61,134 @@ var md5hashes = map[int64]string{
+			volSource: *volSource,
+			testFile:  fmt.Sprintf("%s_io_test_%s", driverInfo.name, f.Namespace.Name),
+			podSec: v1.PodSecurityContext{
+				FSGroup: &fsGroup,
```
Same for all test suites

msau42 commented on test/e2e/storage/subpath.go:
```diff
@@ -168,203 +137,373 @@ var _ = utils.SIGDescribe("Subpath", func() {
+		}
+	case staticPV:
+		framework.Logf("Creating resouce for static PV")
+		if staticPVTestDriver, ok := driver.(staticPVTestDriver); ok {
```
How come the static and dynamic setup cannot reuse what's already done in `genericVolumeTestResource.setup()`?

mkimuram replied:
Subpath test can share generic resource if below two are resolved:
 
1. Subpath test requires ro volume source. For inline volume, readOnly flag is embeded inside specific VolumeSource's struct, so its difficult to override it.
  https://github.com/kubernetes/kubernetes/blob/9e239b1986509458d4dd4896724ea1c703ec0c79/test/e2e/storage/subpath.go#L246

2. Subpath test has pod deletion logic outside It(), so pod is defined in resource to delete there. 
  https://github.com/kubernetes/kubernetes/blob/9e239b1986509458d4dd4896724ea1c703ec0c79/test/e2e/storage/subpath.go#L307

If we skip inline volume test for subpath, 1. will be resolved, so only 2. needs to be considered.

msau42 replied:
Hm is it possible to call `genericVolumeTestResource.setup()` manually, and then do the additional calls needed for 1 + 2 in the `subpath.setup()`?

mkimuram replied:
Oh, I understand your intention, now. I will reflect it. Thank you for your suggestion!

msau42 reviewed (commented):
This is looking nice!  Just some minor comments

mkimuram commented on test/e2e/storage/volume_io.go:
```diff
@@ -61,6 +61,134 @@ var md5hashes = map[int64]string{
+}
+
+func testVolumeIOWithTestInput(testInput *testInput) {
+	Context("-", func() {
```
It starts with:
```
[sig-storage] In-tree Volumes Driver: NFS, Test: volumeIO, TestPattern: Pre-provisioned PV - 
  should write files of various sizes, verify size, validate content [Slow]
```

And ends with:
```
[sig-storage] In-tree Volumes
/root/mkimuram/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/utils/framework.go:22
  Driver: NFS, Test: volumeIO, TestPattern: Pre-provisioned PV
  /root/mkimuram/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/in_tree_volumes.go:92
    -
    /root/mkimuram/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/volume_io.go:138
      should write files of various sizes, verify size, validate content [Slow]
      /root/mkimuram/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/volume_io.go:171
------------------------------
```

msau42 replied:
To make it easier to filter, I think we should have driver in brackets, like `[Driver: nfs]`.

Also, instead of context here being basically empty, can we put the test suite name and test pattern here instead?  That way someone could import this file as a library and have their own runner logic.

msau42 replied:
Right now, most of our volume tests follow the format "<feature> <plugin> <detail>".  Does it make sense to continue that, or do you prefer having plugin first?

mkimuram replied:
Thank you for your suggestion.

```Context``` here is one of the implementations that I'm not satisfied and this is a tricky part to tame ginkgo to work as we expect. I've spent much time to improve it these days, however I couldn't find a good way.

I applyed pattern 2 of "Global Shared Behaviors" described in https://onsi.github.io/ginkgo/#global-shared-behaviors, it requires all the inputs for ```It()``` to be in one struct. Also, to make various ```testSuite```s share the same logic inside the same loop, I made ```testInput``` interface and use type assertion here to convert each ```testInput``` for each ```testSuite```. 

The problem here is that first string argument of ```Context()``` seems to be evaluated, before ```testInput``` is decided, so if we add a code to make a string depending on ```testInput```, it fails with nil pointer dereference. (Also, it is not allowed to pass another arguments to ```test*WithTestInput()```, because of the above reason.)

I can't find a good documentation about it, but I guess that this is due to ginkgo's multistep execution, explained in https://github.com/onsi/ginkgo/issues/462#issuecomment-384633172. From this explanation, we can guess that context string needs to be decided in "Parse stage" to be properly filtered in "Run stage".
(By the same reason, it seems that type assertion needs to be done inside ```BeforeEach```, or in "Run stage", after ```testInput``` is decided.)

Therefore, I think we can put ```testSuite``` dependent string here, because ```testInput``` and ```testSuite``` have 1:1 relationship now. However, we won't be able to put ```testDriver``` or ```testPattern``` dependent string, which will need to be passed as a part of ```testInput```, here.

So, the choices will be:

1. Remove ugly "-" and output ```<plugin> <detail>```, then ```<feature>```
```
[sig-storage] In-tree Volumes
  [Driver: NFS][TestPattern: Pre-provisioned PV]
    [TestSuite: volumeIO]
      should write files of various sizes, verify size, validate content [Slow]
```
2. Prioritize the order of output and keep ugly "-" (We can replace "-" with any static string, though.)
```
[sig-storage] In-tree Volumes
  [TestSuite: volumeIO][Driver: NFS][TestPattern: Pre-provisioned PV]
    -
      should write files of various sizes, verify size, validate content [Slow]
```
3. Continue to search for a better way

msau42 replied:
Thanks for the detailed explanation!

I think the issue like you said, is that `testInput` is evaluated at runtime.  What if you create a different input, like `testMetadata`, which only contains suite name, driver name, and pattern name, and is initialized during Parsing?

mkimuram replied:
Thank you for your suggestion. I tried a similar implementation, but it was not "initialized during Parsing".
I will try this way.

msau42 commented on test/e2e/storage/in_tree_drivers.go:
```diff
@@ -893,9 +892,9 @@ func (g *gceDriver) getDriverInfo() *driverInfo {
 
 func (g *gceDriver) skipUnsupportedTest(testPattern testPattern) {
-	framework.SkipUnlessProviderIs("gce", "gke")
+	framework.SkipUnlessProviderIs("gce", "gke", "custom")
```
ah sorry, "custom" should only be added to nodeOSDistro below, not provider.

msau42 commented on test/e2e/storage/in_tree_volumes.go:
```diff
@@ -373,17 +373,20 @@ func (r *genericVolumeTestResource) setup(driver testDriver, testPattern testPat
 
 	switch testPattern.testVolType {
 	case inlineVolume:
+		framework.Logf("Creating resouce for inline volume")
```
nit "resource"

msau42 commented on test/e2e/storage/subpath.go:
```diff
@@ -220,78 +219,48 @@ func testSubPathWithTestInput(testInput *testInput) {
-				}
-			}
+	case preprovisionedPV:
+		s.genericVolumeTestResource.setup(driver, testPattern)
```
Can this be called before the case statement, since it applies to all 3 cases?

msau42 commented on test/e2e/storage/subpath.go:
```diff
@@ -220,78 +219,48 @@ func testSubPathWithTestInput(testInput *testInput) {
-			}
+	case preprovisionedPV:
+		s.genericVolumeTestResource.setup(driver, testPattern)
+		s.volType = fmt.Sprintf("%sPVC", driverInfo.name)
```
can volType be set to driverInfo.name for all cases?

mkimuram commented on test/e2e/storage/in_tree_volumes.go:
```diff
@@ -0,0 +1,451 @@
+				curTestPattern := testPattern
+				testInput = initTestMetadata(curTestSuiteName, curTestPatternName)
+				curTestInput := testInput
+				testFunc := curTestSuite.getTestFunc(curTestPattern, curDriver)
```
Fixed as suggested. 

Also, in the same commit, I add a code to skip deleting resource and driver when it it is not created by skip. I found that unnecessary PV deletion routine for other clouds is running and taking so much time. It will be worth skipping, although tests won't fail. 

PTAL

msau42 commented on test/e2e/storage/in_tree_volumes.go:
```diff
@@ -0,0 +1,455 @@
+	"aws":                        initAwsDriver,
+}
+
+var testSuites = map[string]func() testSuite{
```
Can this just be a slice without the name of the test suite?  It would be better if the test suite name was defined in the test suite itself (via Context()) instead of in the caller

mkimuram replied:
Changed as suggested.

davidz627 commented on test/e2e/storage/in_tree_volumes.go:
```diff
@@ -0,0 +1,455 @@
+	return volSource, pv, pvc
+}
+
+func (r *genericVolumeTestResource) setup(driver testDriver, testPattern testPattern) {
```
nit: can we use a more descriptive name than "setup". what is this setting up? the driver, the test, the framework?

mkimuram replied:
Renamed as below:

```Setup``` -> ```SetupResource```
```Cleanup``` ->   ```CleanupResource```

mkimuram commented on test/e2e/storage/in_tree_drivers.go:
```diff
@@ -0,0 +1,1187 @@
+func (n *nfsDriver) getVolumeSource(readOnly bool, fsType string) *v1.VolumeSource {
+	return &v1.VolumeSource{
+		NFS: &v1.NFSVolumeSource{
+			Server:   n.serverIP,
```
Thank you for your review.
This error is handled inside New*Server(). So, at the time ```getVolumeSource``` is called it should be set.
Please see https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/volume_util.go#L238

mkimuram commented on test/e2e/storage/in_tree_drivers.go:
```diff
@@ -0,0 +1,1187 @@
+	return &g.driverInfo
+}
+
+func (g *glusterFSDriver) skipUnsupportedTest(testPattern testPattern) {
```
Thank you for pointing it out. I hope that I've fixed all.

mkimuram commented on test/e2e/storage/in_tree_drivers.go:
```diff
@@ -0,0 +1,1187 @@
+}
+
+func (g *glusterFSDriver) getPersistentVolumeSource(readOnly bool, fsType string) *v1.PersistentVolumeSource {
+	// TODO: implement this method propery
```
Implemented all. I hope that it works. (I can't test all of them due to lack of specific provider environments.)

mkimuram commented on test/e2e/storage/in_tree_drivers.go:
```diff
@@ -0,0 +1,1187 @@
+}
+
+func (g *glusterFSDriver) createVolume(testVolType testVolType) {
+	By("deploying GlusterFS driver")
```
This is related to my comment below:

  https://github.com/kubernetes/kubernetes/pull/66577#discussion_r209724412 (see below snip)

As this kind of questions will be asked in the future, I changed my mind and move testServer pod creation logic to ```CreateDriver``` and put explanations that they need to do nothing in ```CreateVolume``` in all these drivers.
Also, for NFS server, switch ```CreateDriver``` logic for inlineVolume/preprovisionedPV or dynamicProvisionedPV.

> For drivers tested with testServer pod, like NFS, iSCSI, and Ceph, volumes are predefined in container images. So, they can't separate ```createDriver``` logic and ```createVolume``` logic.
In addition, dynamic provision pod for NFS seems to be developed separately from that for staticPV/inlineVolume, therefore a different pod from a different container image needs to be launched.  
(This means that two pods, for staticPV/inlineVolume and dynamicPV, will be launched for NFS in dynamic provision test, if we put testServer creation logic in ```createDriver```.)
Therefore, I put testServer creation logic for these drivers in ```createVolume```, not in ```createDriver```.

mkimuram commented on test/e2e/storage/in_tree_drivers.go:
```diff
@@ -0,0 +1,1187 @@
+		ISCSI: &v1.ISCSIVolumeSource{
+			TargetPortal: i.serverIP + ":3260",
+			// from test/images/volumes-tester/iscsi/initiatorname.iscsi
+			IQN:      "iqn.2003-01.org.linux-iscsi.f21.x8664:sn.4b0aae584f7c",
```
It's a constant which is defined below and will not change if we use this pod for tests.
https://github.com/kubernetes/kubernetes/blob/master/test/images/volume/iscsi/initiatorname.iscsi

Actually, The code has been moved from below:
  https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/volume_io.go#L356
  https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/volumes.go#L180

However, path seems to be changed. So, I fixed the path. Thank you for pointing it out.

mkimuram commented on test/e2e/storage/in_tree_volumes.go:
```diff
@@ -0,0 +1,455 @@
+	}
+})
+
+type testDriver interface {
```
I spepareted the file as below:

test/e2e/storage/
├── drivers
│   ├── base.go
│   └── in_tree.go
├── in_tree_volumes.go
├── testpatterns
│   └── testpattern.go
├── testsuites
│   ├── base.go
│   ├── subpath.go
│   ├── volume_io.go
│   ├── volumemode.go
│   └── volumes.go

And put each interface to top of the file below:
  - TestDriver: drivers/base.go
  - TestSuite:   testsuites/base.go
  - TestResource: testsuites/base.go
  - TestInput:    testsuites/base.go

mkimuram commented:
@msau42  @davidz627 

PTAL

This time, I changed a lot of codes, so I would like to review by myself and test locally a bit more.
At the same time, I also would like to have feedback earlier to ensure that we are on the same page.

msau42 commented on test/e2e/storage/testpatterns/testpattern.go:
```diff
@@ -0,0 +1,137 @@
+	Name        string
+	VolType     TestVolType
+	FsType      string
+	VolBindMode storagev1.VolumeBindingMode
```
The test framework itself doesn't seem to be utilizing VolBindMode or VolMode (it's self-contained in the volumemode test suite).  Can we somehow move these to the volumemode file?

msau42 commented on test/e2e/storage/testsuites/volumemode.go:
```diff
@@ -0,0 +1,433 @@
+}
+
+func generateVolumeModeTestFunc(testFunc func(*volumeModeTestInput)) func(*TestInput) {
+	return func(ti *TestInput) {
```
Seems like all test suites have to copy a similar method. Is there anyway we can consolidate this?

mkimuram replied:
@msau42 

Thank you for your suggestion. I also would like to make it a common method, and tried something like below to replace ```generateVolumeModeTestFunc``` to more general one,  but it doesn't work. Because they seem to access invalid memory in parse phase. Any idea or advice for this?

```
func convertToReflectValue(ti *TestInput) reflect.Value {
    var v reflect.Value

    switch t := (*ti).(type) {
    case volumesTestInput:
        v = reflect.ValueOf(&t)
    case volumeIOTestInput:
        v = reflect.ValueOf(&t)
    case subPathTestInput:
        v = reflect.ValueOf(&t)
    case volumeModeTestInput:
        v = reflect.ValueOf(&t)
    default:
        framework.Failf("Fail to convert testInput %v", ti)
    }

    return v
}

func generateTestFunc(testFunc interface{}) func(*TestInput) {
    return func(ti *TestInput) {
        Context((*ti).GetTestMetadataString(), func() {
            var arg reflect.Value

            BeforeEach(func() {
                arg = convertToReflectValue(ti)
            })

            doTest := func(a *reflect.Value) {
                fVal := reflect.ValueOf(testFunc)
                fType := fVal.Type()
                // Check if testFunc is a function that takes exact one argument and return no value
                // ex) func fooTest(*fooTestInput)
                if fType.Kind() != reflect.Func || fType.NumIn() != 1 || fType.NumOut() != 0 {
                    framework.Failf("testFunc %v isn't a suitable function to test", testFunc)
                }

                fVal.Call([]reflect.Value{*a})
            }

            // Call test function
            doTest(&arg)
        })
    }
}
```

msau42 replied:
Hm, does the framework need to pass the input back and forth with the test?  Could the test function take the test pattern as an input, and have a beforeeach() that generates in the input?

mkimuram replied:
Thank you for your advice. 

It seems that I managed to make it work with a bit another approach.
It won't be optimal, but it should improve the readability of the code. 
I will work on other minor fixes and will commit the code for review.

Anyway, thank you so much for your suggestions and advices! The codes become much better.

msau42 commented on test/e2e/storage/in_tree_volumes.go:
```diff
@@ -0,0 +1,207 @@
+	var (
+		cs     clientset.Interface
+		ns     *v1.Namespace
+		node   v1.Node
```
Is this only needed for hostpath tests?  If so, can the hostpath driver setup do this?

msau42 commented on test/e2e/storage/testpatterns/testpattern.go:
```diff
@@ -0,0 +1,137 @@
+)
+
+const (
+	MinFileSize = 1 * framework.MiB
```
These seem specific to the volume_io tests. Can these move there?

mkimuram commented on test/e2e/storage/drivers/base.go:
```diff
@@ -0,0 +1,184 @@
+
+type InlineVolumeTestDriver interface {
+	TestDriver
+	CreateVolume(testpatterns.TestVolType)
```
Fixed as suggested.

mkimuram commented on test/e2e/storage/drivers/base.go:
```diff
@@ -0,0 +1,184 @@
+	}
+}
+
+func startExternalProvisioner(c clientset.Interface, ns string) *v1.Pod {
```
Move to in_tree.go.

mkimuram commented on test/e2e/storage/drivers/base.go:
```diff
@@ -0,0 +1,184 @@
+	IsFsGroupSupported bool
+	IsBlockSupported   bool
+
+	F      *framework.Framework
```
Renamed as suggested.

mkimuram commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-gce-device-plugin-gpu

mkimuram commented:
/test pull-kubernetes-e2e-kops-aws

msau42 commented on test/e2e/framework/volume_util.go:
```diff
@@ -398,10 +398,11 @@ func VolumeTestCleanup(f *Framework, config VolumeTestConfig) {
 		ExpectNoError(err, "Failed to delete client pod: %v", err)
 	}
 
+	if err := f.WaitForPodTerminated(config.Prefix+"-client", ""); !apierrs.IsNotFound(err) {
```
This should actually be waiting for the pod to be deleted.  Terminated only means that the containers have been stopped, but not that the volumes have been unmounted.  Pod deletion indicates that the volumes have been unmounted.

mkimuram replied:
@msau42 

Fixed as suggested. If you meant to remove ```WaitForPodTerminated```, I will do so.

I kept both of them in order to keep maximum time to wait for the status the same. There will be no big difference in actual time to wait by doing so, for these functions will check the status immediately. And will improve the second wait time as it will quit waiting as soon as they find the pod no longer exists.

PTAL

msau42 commented on test/e2e/framework/volume_util.go:
```diff
@@ -398,10 +398,11 @@ func VolumeTestCleanup(f *Framework, config VolumeTestConfig) {
-		}
 		// See issue #24100.
 		// Prevent umount errors by making sure making sure the client pod exits cleanly *before* the volume server pod exits.
 		By("sleeping a bit so kubelet can unmount and detach the volume")
```
After waiting for Pod delete, then this sleep can be removed

mkimuram commented:
/retest

msau42 commented on test/e2e/framework/volume_util.go:
```diff
@@ -68,7 +68,7 @@ const (
 
 	// Waiting period for volume server (Ceph, ...) to initialize itself.
-	VolumeServerPodStartupTimeout = 1 * time.Minute
+	VolumeServerPodStartupTimeout = 3 * time.Minute
```
is this necessary?

mkimuram replied:
In my environment, it takes around 1 minute to set up rbd volume server, as a result, sometime test fails due to the time out. (I hope that my test environment is so much poor.)
Therefore, it would be better to extend it.

msau42 commented on test/e2e/storage/drivers/base.go:
```diff
@@ -0,0 +1,143 @@
+// For example)
+// In gce:
+//  - [Driver: nfs]]
+//  - [Driver: rbd[Feature:Volumes]]
```
I'm not sure if nested tags will mess up the regexes that test-infra uses.  It would be better to not nest them

mkimuram replied:
Fixed as suggested.

msau42 commented on test/e2e/storage/drivers/in_tree.go:
```diff
@@ -0,0 +1,1388 @@
+	case testpatterns.PreprovisionedPV:
+		n.driverInfo.Config, n.serverPod, n.serverIP = framework.NewNFSServer(cs, ns.Name, []string{})
+	case testpatterns.DynamicPV:
+		// TODO(mkimuram): cluster-admin gives too much right but system:persistent-volume-provisioner
```
What is missing from persistent-volume-provisioner?  The existing external provisioner test does not need cluster-admin.

mkimuram replied:
> What is missing from persistent-volume-provisioner? 

Test fails with below error:

```
      Failed to create injector pod: an empty namespace may not be set during creation
      Expected error:
          <*errors.errorString | 0xc421aa7600>: {
              s: "an empty namespace may not be set during creation",
          }
          an empty namespace may not be set during creation
      not to have occurred
```

And system:persistent-volume-provisioner in my test environment is as follows:
```
# kubectl describe clusterrole system:persistent-volume-provisioner
Name:         system:persistent-volume-provisioner
Labels:       kubernetes.io/bootstrapping=rbac-defaults
Annotations:  rbac.authorization.kubernetes.io/autoupdate=true
PolicyRule:
  Resources                      Non-Resource URLs  Resource Names  Verbs
  ---------                      -----------------  --------------  -----
  persistentvolumes              []                 []              [create delete get list watch]
  persistentvolumeclaims         []                 []              [get list update watch]
  storageclasses.storage.k8s.io  []                 []              [get list watch]
  events                         []                 []              [watch create patch update]
```

While nfs provisioner seems to require more:

https://github.com/kubernetes-incubator/external-storage/blob/master/nfs-client/deploy/auth/clusterrole.yaml

> The existing external provisioner test does not need cluster-admin.

I guess that this issue doesn't appear until the test try to create a pod which attaches the provisioned volume.

I intend to fix it after this PR is merged. Because it is beyond scope of refactoring.
(My intention for adding NFS here was to test dynamic provision code in my test environment, for it was only dynamic provisioner that can be tested without cloud provider.)

msau42 commented on test/e2e/storage/drivers/in_tree.go:
```diff
@@ -0,0 +1,1388 @@
+}
+
+func (g *glusterFSDriver) SkipUnsupportedTest(tp testpatterns.TestPattern) {
+	framework.SkipUnlessNodeOSDistroIs("gci", "ubuntu", "custom")
```
I think this first skip can be removed.  All gce/gke suites should now be only those 3.

mkimuram replied:
I think that original intention of this skip was to skip gluster test on a OS distro that doesn't support glusterfs. For debian or fedora base OS distro, it would be needed. And it won't be cloud provider independent factor.
(They don't have a CI system, so removing this wouldn't impact any CI, though.)

https://github.com/kubernetes/kubernetes/commit/ecca32641f9dd3143da94b08eeb7f3d33fb458a3#diff-c83ebba11ffd12bd594ac8aa2f9301ffR143

msau42 commented on test/e2e/storage/drivers/in_tree.go:
```diff
@@ -0,0 +1,1388 @@
+func (g *glusterFSDriver) SkipUnsupportedTest(tp testpatterns.TestPattern) {
+	framework.SkipUnlessNodeOSDistroIs("gci", "ubuntu", "custom")
+	if tp.FsType == "xfs" {
+		framework.SkipUnlessNodeOSDistroIs("ubuntu", "custom")
```
@gnufied do you guys run the gluster test in a separate CI?  If so, then I think this Skip can be conditioned on gce/gke provider.

msau42 commented on test/e2e/storage/in_tree_volumes.go:
```diff
@@ -0,0 +1,228 @@
+
+						// Create driver
+						vt = tp.VolType
+						drv.CreateDriver(vt)
```
Can Create/Cleanup Driver be called in the outermost loop?  I am thinking in the case of CSI, you may bring up the cluster and the driver once in the beginning, and then run all the test cases against it, and then bring down the driver/cluster at the end.

mkimuram replied:
I agree with you that it will be much cleaner. However, existing containerized volume test server, such as iSCSI and rbd, have a predefined volume inside. And, I think that existing tests are implemented in a way that volume, or the containerized volume test server, is expected to be initialized before each test. To move these test as they are, create/cleanup will be needed to be done for each test.

I think that we would be better improving this in the next step, after we achieve that all containerized volume test servers implement actual ```createVolume```/```deleteVolume```.

msau42 replied:
Can the iscsi and rbd servers be moved to CreateVolume instead?

msau42 commented on test/e2e/storage/in_tree_volumes.go:
```diff
@@ -0,0 +1,228 @@
+	for _, initDriver := range testDrivers {
+		curDriver := initDriver()
+
+		for _, testSuiteInit := range testSuites {
```
Can this be enclosed in a function called something like "runTestSuite"?  I want to separate out the driver setup from the test cases to support the scenario where the driver may have already been preinstalled or setup.

msau42 commented on test/e2e/storage/testsuites/base.go:
```diff
@@ -0,0 +1,319 @@
+			var ok bool
+
+			switch fn := testFunc.(type) {
+			case func(*volumesTestInput):
```
This is leaking test suite implementation details into the framework.  Is there a way we can avoid having a TestInput at all?

For example, could `testSubpath` include the Context() and BeforeEach() that it needs to setup all the necessary inputs (including the genericTestResource) ?  Then in the framework, you don't need a CreateTestInput method.

mkimuram replied:
Could you explain a little bit more about your idea?
 
I still don't understand your idea completely, but I'm worry about that it would require all TestSuite implement the similar codes. Also, to leave testSubPath() accept testInput will allow other codes reuse testSubPath(), which I guess would still benefit some of the existing tests.
(I agree with you that current implementation is not optimal, though.)

msau42 replied:
Something like:

```
func testSubpath(driver, testPattern) {
  Context("Subpath") , func() {
     BeforeEach(func() {
          // Setup all the test inputs and resources here
    })

     After Each(func() {
         // Cleanup
    }

   It("test case 1")

   It("test case 2")

  }
}

davidz627 commented on test/e2e/storage/drivers/base.go:
```diff
@@ -0,0 +1,143 @@
+	FeatureTag string // FeatureTag for the driver
+
+	MaxFileSize        int64           // Max file size to be tested for this driver
+	SupportedFsType    map[string]bool // Map of string for supported fs type
```
nit: Instead of `map[string]bool` can we use `sets.String` instead? Its the apimachinery type that pretends to be a `set` since go doesn't have them.

mkimuram replied:
@davidz627 

Thank you for your pointing it out. I will fix it as suggested.

davidz627 commented on test/e2e/storage/testsuites/base.go:
```diff
@@ -0,0 +1,330 @@
+}
+
+// TestInput represents an interface for an input to be passed for the test function
+type TestInput interface {
```
this seems like a strange interface to me, could you explain why it is necessary?

mkimuram replied:
It might be O.K just to make ```prepareTest()``` to accept ```interface{}``` and do type assertion from ```interface{}```, not from ```TestInput```.
My intention was to ensure that all ```*TestInput``` is to use for ```TestInput```.
I'm new to golang and lack of this kind of knowledge. So I will fix as so, if it is better to implement this way in this case.

davidz627 commented on test/e2e/storage/testsuites/base.go:
```diff
@@ -0,0 +1,330 @@
+// is not suitable to be tested.
+// Whether it needs to be skipped is checked by following steps:
+// 1. Check if Whether volType is supported by driver from its interface
+// 2. Check if fsType is supporetd by driver
```
nit: supported

davidz627 commented on test/e2e/storage/testsuites/subpath.go:
```diff
@@ -0,0 +1,755 @@
+	mountImage      = imageutils.GetE2EImage(imageutils.Mounttest)
+)
+
+type subPathTestSuite struct {
```
for each of the types in your change, could you explicitly specify which interfaces it implements. For example:
`var _ TestSuite = &subPathTestSuite{}`
This forces compile time checking of interface implementation and makes it explicit for any reader to know what interfaces this type implements

mkimuram replied:
Fixed as suggested. PTAL

mkimuram commented:
@msau42 

I moved driver creation logic to outer loop as suggested. PTAL

Note that driver creation codes are still executed before each tests, even though it is moved to ```BeforeEach``` in out most loop. If there are any good ways to make them execute once before all test executions, please let me know them. I will apply to the current code.

Also, due to the above reason, nfs tests currently deploy an external provisioner every time even when it is not dynamic provisioned PV tests, because deploying external provisioner is done in driver creation in out most loop and it doesn't know whether it is for dynamic provision at the time of this loop. It won't be optimal until above improvement is done.

davidz627 commented on test/e2e/storage/testsuites/volume_io.go:
```diff
@@ -0,0 +1,370 @@
+// Test files are generated in testVolumeIO()
+// If test file generation algorithm changes, these must be recomputed.
+var md5hashes = map[int64]string{
+	testpatterns.FileSizeSmall:  "5c34c2813223a7ca05a3c2f38c0d1710",
```
could these be computed based on the test file generation instead of hardcoded?

mkimuram replied:
Actually, I've just move it from volume_io.go defined in test/e2e/storage, and I would like to make this kind of improvement out of the scope of this refactoring. 
We would be able to improve it in a separate commit.

davidz627 commented on test/e2e/storage/testsuites/volume_io.go:
```diff
@@ -0,0 +1,370 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
A general question: have you timed runs of these tests vs. the old ones? are there any significant performance increases/decreases? Are we running the exact same set of tests (at least the same number) now as before

msau42 replied:
This refactoring will run more tests than before.  Which is good.  Previously, you needed to create a test case for every plugin.  This refactoring makes it easier to run any test case against any plugin.

davidz627 reviewed (commented):
looking really good! Thanks for staying on top of this

mkimuram commented:
@msau42 

As for ```execTest()```, which are introduced this time, similar codes are implemented in each ```TestSuite``` again, although duplicated codes are minimized by calling TestSuite's ```prepareTest()``` and ```teardown()```.

We might be able to reduce the duplicated code, by applying the same way that was done in ```generateTestFunc()``` https://github.com/kubernetes/kubernetes/pull/66577/commits/79891d0cfcc114c6f5d24e8b1c3b257572fb42e0#diff-5a0e6c4939cdb1706765e89ecae21f00L95

And this will leak the framework implementation again, so we will need to balance them.
I guess that you prefer current implementation from your previous comment.

msau42 commented on test/e2e/storage/testsuites/volume_io.go:
```diff
@@ -0,0 +1,368 @@
+
+		BeforeEach(func() {
+			tr, ti = prepareTest(drv, t, tp)
+			if vti, ok = (ti).(volumeIOTestInput); !ok {
```
I like this a lot better than before, but I'm wondering if it's really necessary to do the test input initialization in a common function and have to do this object casting.

Can we just call the common setupResoource here and then create the volumeIOTestInput directly?

mkimuram replied:
I understand your intention. I will separate resource initialization logic and testInput creation logic and only make former one to the common function.

msau42 commented on test/e2e/storage/testsuites/base.go:
```diff
@@ -0,0 +1,326 @@
+	skipUnsupportedTest(ts, drv, tp)
+
+	// Create volume for pre-provisioned volume tests
+	drivers.CreateVolume(drv, vt)
```
Can this go inside setupResource?

mkimuram replied:
@msau42 

Fixed as suggested. PTAL

msau42 commented:
> Note that driver creation codes are still executed before each tests

Oh that's interesting, not what I expected.  I still prefer this way though, it will make it easier to run the test suites on a platform that may already have all the drivers installed.

mkimuram commented:
@msau42 

> Oh that's interesting, not what I expected. I still prefer this way though, it will make it easier to run the test suites on a platform that may already have all the drivers installed.

Thank you for you comment. 
Then, I will continue to refactor on top of the current commit and will ask you for further review.

mkimuram commented:
/retest

mkimuram commented:
/retest

msau42 commented on test/e2e/storage/drivers/base.go:
```diff
@@ -0,0 +1,176 @@
+func GetDriverNameWithFeatureTags(drv TestDriver) string {
+	di := drv.GetDriverInfo()
+
+	return fmt.Sprintf("Driver: %s%s ", di.Name, di.FeatureTag)
```
Can you put the driver into bracket? Like `[Driver: cephfs] [Feature: blah]`

msau42 commented on test/e2e/storage/testsuites/base.go:
```diff
@@ -0,0 +1,320 @@
+}
+
+func getTestNameStr(ts TestSuite, tp testpatterns.TestPattern) string {
+	return fmt.Sprintf("Testsuite: %s Testpattern: %s", ts.getName(), tp.Name)
```
Can you put these in brackets too? `[TestSuite: %s] [TestPattern: %s]`

msau42 commented on test/e2e/storage/testsuites/volume_io.go:
```diff
@@ -0,0 +1,364 @@
+func (t *volumeIOTestSuite) skipUnsupportedTest(tp testpatterns.TestPattern, drv drivers.TestDriver) {
+}
+
+func (t *volumeIOTestSuite) initTestResource() TestResource {
```
I'm not convinced that we need to separate out resource creation from test input creation.  Can createVolumeIOTestInput just call the genericVolumeTestResource.setupResource()?

mkimuram replied:
Fixed as suggested. PTAL

mkimuram commented:
/retest

mkimuram commented:
@msau42 

This is out of the scope of this refactoring, but I often see the below subpath test with nfs fails with a below error:

[Test]
```[sig-storage] In-tree Volumes [Driver: nfs] [Testsuite: subPath][Testpattern: Inline-volume (default fs)] should support existing directories when readOnly specified in the volumeSource```

[Error]
```
STEP: Creating a pod to test subpath
Aug 25 02:48:25.397: INFO: Waiting up to 5m0s for pod "pod-subpath-test-nfs-btj7" in namespace "e2e-tests-volumes-pdlpg" to be "success or failure"
Aug 25 02:48:25.448: INFO: Pod "pod-subpath-test-nfs-btj7": Phase="Pending", Reason="", readiness=false. Elapsed: 51.363029ms
Aug 25 02:48:27.509: INFO: Pod "pod-subpath-test-nfs-btj7": Phase="Pending", Reason="", readiness=false. Elapsed: 2.111631329s
Aug 25 02:48:29.561: INFO: Pod "pod-subpath-test-nfs-btj7": Phase="Running", Reason="", readiness=true. Elapsed: 4.163580676s
Aug 25 02:48:31.612: INFO: Pod "pod-subpath-test-nfs-btj7": Phase="Running", Reason="", readiness=true. Elapsed: 6.215255458s
Aug 25 02:48:33.664: INFO: Pod "pod-subpath-test-nfs-btj7": Phase="Running", Reason="", readiness=true. Elapsed: 8.267138552s
Aug 25 02:48:35.716: INFO: Pod "pod-subpath-test-nfs-btj7": Phase="Running", Reason="", readiness=true. Elapsed: 10.318755036s
Aug 25 02:48:37.769: INFO: Pod "pod-subpath-test-nfs-btj7": Phase="Running", Reason="", readiness=true. Elapsed: 12.371954726s
Aug 25 02:48:39.821: INFO: Pod "pod-subpath-test-nfs-btj7": Phase="Running", Reason="", readiness=true. Elapsed: 14.423758437s
Aug 25 02:48:41.873: INFO: Pod "pod-subpath-test-nfs-btj7": Phase="Running", Reason="", readiness=false. Elapsed: 16.475540254s
Aug 25 02:48:43.926: INFO: Pod "pod-subpath-test-nfs-btj7": Phase="Failed", Reason="", readiness=false. Elapsed: 18.528621557s
Aug 25 02:48:44.033: INFO: Output of node "ip-172-20-59-177.us-west-2.compute.internal" pod "pod-subpath-test-nfs-btj7" container "test-container-subpath-nfs-btj7": Error reading file /test-volume/test-file: open /test-volume/test-file: no such file or directory, retrying

Aug 25 02:48:44.091: INFO: Output of node "ip-172-20-59-177.us-west-2.compute.internal" pod "pod-subpath-test-nfs-btj7" container "test-container-volume-nfs-btj7": content of file "/test-volume/e2e-tests-volumes-pdlpg/test-file": mount-tester new file

mode of file "/test-volume/e2e-tests-volumes-pdlpg/test-file": -rw-r--r--
```

Is my understanding right that [```testFileContent```](https://github.com/kubernetes/kubernetes/blob/master/test/images/mounttest/mt.go#L266) should return nil instead of break in the loop?
Because it should allow success in the second or later attempts, but current code returns non-nil if the first attempt fails.  I will open another issue and make a PR.

Note that this error seems to happen often with nfs, because nfs is mounted with async option by default, which will make read access happen before write completes.

msau42 commented:
@mkimuram I'm not sure I see the issue.  testFileContent should retry multiple times if it failed to read the file or if the file content didn't match.  Only when it times out, then it will return error.  You can see from the log above that the pod ran for some 15 seconds before erroring.  Maybe the timeout needs to be increased?

msau42 commented:
Sorry I keep changing my mind on the test naming format. I think something like this would be easier to read and understand what is being tested
```
[sig-storage] In-tree Volumes [Driver: nfs] [Testpattern: Inline-volume (default fs)] Subpath should support existing directories when readOnly specified in the volumeSource
```

mkimuram commented:
> Sorry I keep changing my mind on the test naming format. I think something like this would be easier to read and understand what is being tested
> ```[sig-storage] In-tree Volumes [Driver: nfs] [Testpattern: Inline-volume (default fs)] Subpath should support existing directories when readOnly specified in the volumeSource```

I prefer former one, so I will change as so.

mkimuram commented:
@msau42 

>  You can see from the log above that the pod ran for some 15 seconds before erroring. Maybe the timeout needs to be increased?
Thank you for your comment. I was wrong.

I checked the above log again and found that there were outputs from [```MatchContainerOutput```](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L2399) and [```waitForPodSuccessInNamespaceTimeout```](https://github.com/kubernetes/kubernetes/blob/52994701526a779d6a19faec66f0ca8fb064c7c6/test/e2e/framework/util.go#L1586), so the error should be due to that the pod status went to PodFailed phase. And this happens before it reaches timeout. I'm still not sure why it happpens, though. Any ideas?

Note that entire logs that I was checking were [this](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/66577/pull-kubernetes-e2e-kops-aws/103173/#sig-storage-in-tree-volumes-driver-nfs-testsuite-subpathtestpattern-inline-volume-default-fs-should-support-existing-directories-when-readonly-specified-in-the-volumesource)

msau42 commented:
Regarding nfs test failure, you only see this sometimes?  The retry duration in the subpath test is set to 10 seconds.  Maybe that is too short and should be increased?

mkimuram commented:
/retest

mkimuram commented:
> Regarding nfs test failure, you only see this sometimes? The retry duration in the subpath test is set to 10 seconds. Maybe that is too short and should be increased?

Thank you for your comment. 

I'm not sure why but it only happens in aws environment and it happens quite often.
(It doesn't happen in GCE and my local environment.)
Let me try increasing retryDuration to 20 and see if it solves this issue.

mkimuram commented:
/test pull-kubernetes-e2e-kops-aws

mkimuram commented:
@msau42 

After retryDuration was increased, no failure happened to subpath tests in AWS. 
(Although, there was a timeout error in the first run.)

PTAL

msau42 commented on test/e2e/storage/testsuites/subpath.go:
```diff
@@ -0,0 +1,746 @@
+func (s *subPathTestSuite) execTest(drv drivers.TestDriver, tp testpatterns.TestPattern) {
+	Context(getTestNameStr(s, tp), func() {
+		var (
+			tr           subPathTestResource
```
Can we use non-abbreviated variable names throughout?  Like "resource", "input", and "pattern"?  It's more verbose, but will be easier to read and understand.

msau42 commented:
I think we can squash the commits now.  I'm largely ok with this change.

mkimuram commented:
/retest

mkimuram commented:
/retest

mkimuram commented:
/retest

mkimuram commented:
/retest

mkimuram commented:
@msau42 

I renamed abbreviated variables and squashed the commits.
Also, I made additional two minor changes to avoid test failures. PTAL
(I made them two different commits to make it easier to review. I will squash the commits again, once you confirm that these changes make sense.)

msau42 commented:
lgtm! i'll wait for your squash

mkimuram commented:
Thank you for your review. I squashed the commits again.

msau42 commented:
/lgtm

msau42 commented:
/assign @saad-ali 
for test approval

/assign @vishh 
for hack approval

vishh commented:
/approve

jsafrane commented:
Trying 1.12, it is related to tests only and it helps us to increase test coverage.

mkimuram committed:
Refactor volume test in a similar way to csi tests

Refactoring for non-csi e2e test similar to below commit in csi e2e test.
https://github.com/kubernetes/kubernetes/commit/4d11dab272fc358c918a1a6b83db8f938535cdb0#diff-0d9ecaa3e6a0297186ad33f57aad472e

Scopes for this refactoring are below four files:
  - test/e2e/storage/volumes.go
  - test/e2e/storage/volume_io.go
  - test/e2e/storage/persistent_volumes-volumemode.go
  - test/e2e/storage/subpath.go

fixes: #66571

mkimuram commented:
I rebased to upstream/master again and modified the commit date.

mkimuram commented:
/retest

mkimuram commented:
/test pull-kubernetes-e2e-kops-aws

msau42 commented:
/lgtm

saad-ali commented:
/milestone v1.12

msau42 commented:
That last failure had many of the new nfs tests error.  Is this a new flake that we could be introducing?

msau42 commented:
Many slow tests are failing now.  @mkimuram , can you take a look?

msau42 commented:
https://k8s-testgrid.appspot.com/sig-storage#gce-slow

mkimuram commented:
@msau42 

It seems that failing tests are below two and both tests call ```testPodContainerRestart```
- subPath should support restarting containers using directory as subpath
- subPath should support restarting containers using file as subpath

As seen in the logs, subpath pod deletion that should be called from ```cleanupResource``` isn't called. (Only test server pod is deleted there.) As a result, namespace deletion after the test fails due to the pod remaining in the namespace.
So, the quick workaround would be to add  defer func() { framework.DeletePodWithWait(f, f.ClientSet, pod)}() to ```testPodContainerRestart```, as it is done in most of the other subPath tests. Then, we should investigate more on ```subpathTestResource```'s ```cleanupResource```.

```
STEP: Waiting for container to stop restarting
Aug 30 08:11:33.142: INFO: Container has restart count: 2
Aug 30 08:12:35.132: INFO: Container restart has stabilized
[AfterEach] [Testpattern: Inline-volume (default fs)] subPath
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/subpath.go:119
STEP: Deleting pod
Aug 30 08:12:35.132: INFO: Deleting Gluster endpoints "gluster-server"...
Aug 30 08:12:35.244: INFO: Deleting Gluster server pod "gluster-server"...
Aug 30 08:12:35.244: INFO: Deleting pod "gluster-server" in namespace "e2e-tests-volumes-zhmp8"
Aug 30 08:12:35.356: INFO: Wait up to 5m0s for pod "gluster-server" to be fully deleted
```
